File: WINGS_AI_Bootcamp/Python/Pyramid1_diamond1.py
# ASOKAN

STAR, SPACE = "*", " "
LF = "\n"
START, REPEAT, STOP = STAR, STAR + STAR, ''
WIDTH = 60

def pattern(size: int) -> list[str]:
    return [line(i) for i in range(size)]

def line(k: int) -> str:
    return start(START, k) + repeat(REPEAT, k) + stop(STOP, k)

def start(atom: str, k: int) -> str:
    return atom

def repeat(atom: str, k: int) -> str:
    return k * atom

def stop(atom: str, k: int) -> str:
    return atom

def make_pyramid(size: int) -> str:
    return LF.join(line.center(WIDTH) for line in pattern(size))

def make_diamond(size: int) -> str:
    half = size // 2 if size % 2 == 0 else (size + 1) // 2
    diamond = pattern(half) + pattern(size - half)[::-1]
    return LF.join(line.center(WIDTH) for line in diamond)

print(make_pyramid(7))
print(make_diamond(8))

File: WINGS_AI_Bootcamp/Haskell/Pyramid1_diamond1.hs
import Data.List (intercalate)

width :: Int
width = 60

start :: Int -> String
start _ = "*"

repeatStar :: Int -> String
repeatStar k = concat $ replicate k "**"

stop :: Int -> String
stop _ = ""

line :: Int -> String
line k = start k ++ repeatStar k ++ stop k

pattern :: Int -> [String]
pattern size = map line [0..size-1]

center :: Int -> String -> String
center w s = replicate padding ' ' ++ s
  where padding = (w - length s) `div` 2

makePyramid :: Int -> String
makePyramid size = intercalate "\n" $ map (center width) (pattern size)

makeDiamond :: Int -> String
makeDiamond size =
  let half = if even size then size `div` 2 else (size + 1) `div` 2
      top = pattern half
      bottom = reverse (pattern (size - half))
  in intercalate "\n" $ map (center width) (top ++ bottom)

main :: IO ()
main = do
  putStrLn $ makePyramid 7
  putStrLn $ makeDiamond 8

